rust/src/crypto.rs

```diff
@@ -0,0 +1,314 @@
+use derive_more::{AsMut, AsRef, From};
+use sodiumoxide::crypto::{box_, sealedbox, sign};
+
+pub trait ByteObject: Sized {
+    /// Create a new object with all the bytes initialized to `0`.
+    fn zeroed() -> Self;
+
+    /// Get the object byte representation
+    fn as_slice(&self) -> &[u8];
+
+    /// Create a object from the given buffer. This function will fail
+    /// and return `None` if the length of the byte-slice isn't equal to
+    /// the length of the object.
+    fn from_slice(bytes: &[u8]) -> Option<Self>;
+
+    /// Create a object from the given buffer.
+    ///
+    /// # Panic
+    ///
+    /// This function will panic if the length of the byte-slice isn't
+    /// equal to the length of the object.
+    fn from_slice_unchecked(bytes: &[u8]) -> Self {
+        Self::from_slice(bytes).unwrap()
+    }
+}
+
+/// Generate a new random key pair
+pub fn generate_encrypt_key_pair() -> (PublicEncryptKey, SecretEncryptKey) {
+    let (pk, sk) = box_::gen_keypair();
+    (PublicEncryptKey(pk), SecretEncryptKey(sk))
+}
+
+/// Deterministically derive a new key pair from the given seed
+pub fn derive_key_pair_from_seed(seed: &KeySeed) -> (PublicEncryptKey, SecretEncryptKey) {
+    let (pk, sk) = box_::keypair_from_seed(seed.as_ref());
+    (PublicEncryptKey(pk), SecretEncryptKey(sk))
+}
+
+/// A seed that can be used for key pair generation. When `KeySeed`
+/// goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct KeySeed(box_::Seed);
+
+impl ByteObject for KeySeed {
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::Seed::from_slice(bytes).map(Self)
+    }
+
+    fn zeroed() -> Self {
+        Self(box_::Seed([0; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// A seed that can be used for signing key pair generation. When
+/// `KeySeed` goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct SignKeySeed(sign::Seed);
+
+/// Generate a new random signing key pair
+pub fn generate_sign_key_pair() -> (PublicSigningKey, SecretSigningKey) {
+    let (pk, sk) = sign::gen_keypair();
+    (PublicSigningKey(pk), SecretSigningKey(sk))
+}
+
+/// Deterministically derive a new signing key pair from the given
+/// seed
+pub fn derive_sign_key_pair_from_seed(seed: &SignKeySeed) -> (PublicSigningKey, SecretSigningKey) {
```

Comment:
I picked "derive" here because I wanted to highlight that unlike `generate_key_pair` above, this function was deterministic: given the same input it will always output the same thing. I don't know if it really matters though.

Reply:
on one hand the deterministic nature of this function is already implicitly given by `..._from_seed()`, on the other hand it probably doesn't do any harm to have a `derive_...` as an explicit action. no strong preference for either one.

Reply:
Note that the `_from_seed()` suffix is gone now that it's a method of `EncryptKeySeed`.

@finiteprods what do you think of the arguments above?